train/NEAT/evolve.py

```diff
@@ -0,0 +1,157 @@
+import math
+import os
+import sys
+import time
+import numpy as np
+import pickle
+import argparse
+
+import neat
+import gym
+import ppaquette_gym_super_mario
+from ppaquette_gym_super_mario.wrappers import *
+
+import visualise
+
+# -----------------------------------------------------------------------------
+#  GLOBALS
+# -----------------------------------------------------------------------------
+args = None
+env = []
+
+for i in range(0, 32):
+    env.append(gym.make('ppaquette/SuperMarioBros-{:d}-{:d}-Tiles-v0'.format(int((i/4)+1), int((i%4)+1))))
+
+# -----------------------------------------------------------------------------
+#  NEAT TRAINING
+# -----------------------------------------------------------------------------
+
+# Creates valid button presses to pass to gym
+def clean_outputs(outputs):
+    return [1 if sigmoid(b) > 0.5 else 0 for b in outputs]
+
+# Simple sigmoid function. 
+def sigmoid(x):
+    return (1 / (1 + np.exp(-x)))
+
+#TODO: Need a better fitness function.
+def calculate_fitness(info):
+    return info['distance']
+
+def eval_genome(genome, config):
+
+    net = neat.nn.FeedForwardNetwork.create(genome, config)
+    stuckMax = 600
+    
+    for i in range(0, 32):
+        observation = env[i].reset()
+        done = False
+        stuck = 0
+        while not done:
+            # Get move from NN
+            outputs = clean_outputs(net.activate(observation.flatten()))
+
+            # Make move
+            observation, reward, done, info = env[i].step(outputs)
+
+            # Check if Mario is progressing in level
+            stuck += 1 if reward <= 0 else 0
+
+            #TODO: Needs improvment, need to disable at end of level and when in a pipe.
+            if (stuck > stuckMax):
+                env[i].close()
+                return calculate_fitness(info)
+
+        if info['life'] == 0:
+            break
+
+        env[i].close()
+    return calculate_fitness(info)
+
+def run(config):
+    if args.checkpoint:
+        pop = neat.Checkpointer.restore_checkpoint(args.checkpoint)
+    else:
+        pop = neat.Population(config)
+
+    # Checkpoint every 5 generations or 10 min.
+    pop.add_reporter(neat.Checkpointer(5, 600))
+
+    pop.add_reporter(neat.StdOutReporter(True))
+    stats = neat.StatisticsReporter()
+    pop.add_reporter(stats)
+
+    pe = neat.ParallelEvaluator(args.numCores, eval_genome)
+    
+    # We could use this to set a limit of generations... But i dont think we need too... 
+    #for x in range(args.generations if args.generations > 0 else sys.maxsize): 
+    while True:
+        winner = pop.run(pe.evaluate, 5)
+        
+        if args.vineLogging:
+            #TODO: Fix Vine Logging, everything needed should be in stats. 
+            visualise.save_statistics(stats)
```

Comment:
Presumably the vine logging function just needs a new way of working?

Reply:
Before you were saving things to a file, passing a whole lot of stuff. I am sure that everything you need is in stats. I was hoping that someone who already knows how the vine logging works can jump in and fix it. if you want i can go over how vine logging works, but it just seems like a waste of time as i think Jason could do it in a fraction of the time as he is already familiar with all that vine logging code given that he wrote it originally.

Reply:
I rewrote the vine logging for the multi-threading so happy to port it again 👍